File: Codility/04. Counting Elements/MaxCounters/Solution.swift
// https://app.codility.com/programmers/lessons/4-counting_elements/max_counters/
// Respectable

public func solution(_ N : Int, _ A : inout [Int]) -> [Int] {
    let M = A.count
    var R = [Int](repeating: 0, count: N)

    var MAX_VALUE = 0
    var LAST_MAX_VALUE = 0

    for I in 0..<M {
        if A[I] <= N {
            R[A[I] - 1] = max(R[A[I] - 1], LAST_MAX_VALUE)
            R[A[I] - 1] += 1
            MAX_VALUE = max(MAX_VALUE, R[A[I] - 1])
        } else {
            LAST_MAX_VALUE = MAX_VALUE
        }
    }

    if LAST_MAX_VALUE > 0 {
        for I in 0..<N {
            R[I] = max(R[I], LAST_MAX_VALUE)
        }
    }

    return R
}

File: Codility/04. Counting Elements/MaxCounters/Solution.py
# https://app.codility.com/programmers/lessons/4-counting_elements/max_counters/
# Respectable

from typing import List


def solution(N: int, A: List[int]) -> List[int]:
    """Apply increments directly and defer max-counter ops via a lazy floor raised to the running max."""
    # Time: O(n + N)   Space: O(N)
    counters = [0] * N
    current_max = 0
    floor = 0  # value all counters are lazily raised to on a max-counter op

    for op in A:
        if op <= N:
            idx = op - 1
            counters[idx] = max(counters[idx], floor) + 1
            current_max = max(current_max, counters[idx])
        else:
            floor = current_max

    return [max(c, floor) for c in counters]


def test() -> None:
    assert solution(5, [3, 4, 4, 6, 1, 4, 4]) == [3, 2, 2, 4, 2]
    assert solution(3, [4, 4, 4]) == [0, 0, 0]
    assert solution(2, [1, 1, 2]) == [2, 1]
